File: backend/app/api/routes/contracts.py
"""Contracts routes — upload, list, status, download assets."""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, File, HTTPException, Query, UploadFile, status
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

from app.api.deps import CurrentUser, DbSession, require
from app.api.services import local_storage
from app.api.services.document_parser import parse_document
from app.core.config import settings
from app.core.logging import get_logger
from app.models.audit_log import write_audit_event
from app.models.contract import Contract
from app.workers.tasks import run_review_pipeline

logger = get_logger(__name__)
router = APIRouter(prefix="/contracts", tags=["contracts"])


class ContractOut(BaseModel):
    id: str
    name: str
    status: str
    overall_risk_score: int
    clause_count: int
    thread_id: str

    class Config:
        from_attributes = True


@router.post("/upload", response_model=ContractOut, status_code=status.HTTP_201_CREATED)
async def upload_contract(
    file: Annotated[UploadFile, File(description="PDF or DOCX contract")],
    user: CurrentUser,
    db: DbSession,
):
    if file.size and file.size > settings.MAX_UPLOAD_MB * 1024 * 1024:
        raise HTTPException(413, detail=f"File exceeds {settings.MAX_UPLOAD_MB} MB limit")

    content = await file.read()
    contract_id = str(uuid.uuid4())

    try:
        raw_text = parse_document(content, file.filename or "contract.pdf")
    except Exception as exc:
        raise HTTPException(422, detail=f"Could not parse document: {exc}")

    storage_key = f"contracts/{user.id}/{contract_id}/{file.filename}"
    try:
        local_storage.put(content, storage_key, file.content_type or "application/octet-stream")
    except Exception as exc:
        logger.error("local_storage_upload_failed", key=storage_key, error=str(exc))
        raise HTTPException(500, detail="Could not save uploaded file")

    contract = Contract(
        id=contract_id,
        owner_id=user.id,
        name=file.filename or "Unnamed Contract",
        original_filename=file.filename or "",
        file_type=(file.filename or "").rsplit(".", 1)[-1].lower(),
        storage_key=storage_key,
        raw_text=raw_text,
        status="pending",
    )
    db.add(contract)

    write_audit_event(
        db,
        contract_id=contract_id,
        event_type="contract_uploaded",
        user_id=user.id,
        payload={"filename": file.filename, "size_bytes": len(content)},
    )
    db.commit()
    db.refresh(contract)

    run_review_pipeline.delay(contract_id, raw_text, file.filename or "", user.id)
    logger.info("contract_queued", contract_id=contract_id)

    return contract


@router.get("/", response_model=list[ContractOut])
def list_contracts(
    user: CurrentUser,
    db: DbSession,
    skip: int = Query(0, ge=0),
    limit: int = Query(20, le=100),
):
    return (
        db.query(Contract)
        .filter(Contract.owner_id == user.id)
        .order_by(Contract.created_at.desc())
        .offset(skip)
        .limit(limit)
        .all()
    )


@router.get("/{contract_id}", response_model=ContractOut)
def get_contract(contract_id: str, user: CurrentUser, db: DbSession):
    contract = _get_or_404(db, contract_id, user.id)
    return contract


@router.get("/{contract_id}/download/redline")
def download_redline(contract_id: str, user: CurrentUser, db: DbSession):
    contract = _get_or_404(db, contract_id, user.id)
    if not contract.review or not contract.review.redlined_docx_key:
        raise HTTPException(404, detail="Redlined DOCX not yet generated")

    try:
        stream = local_storage.get(contract.review.redlined_docx_key)
    except FileNotFoundError:
        raise HTTPException(404, detail="Redlined DOCX file is missing from storage")

    return StreamingResponse(
        stream,
        media_type="application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        headers={"Content-Disposition": f'attachment; filename="redlined_{contract.name}.docx"'},
    )


@router.get("/{contract_id}/download/report")
def download_report(contract_id: str, user: CurrentUser, db: DbSession):
    contract = _get_or_404(db, contract_id, user.id)
    if not contract.review or not contract.review.risk_pdf_key:
        raise HTTPException(404, detail="Risk report PDF not yet generated")

    try:
        stream = local_storage.get(contract.review.risk_pdf_key)
    except FileNotFoundError:
        raise HTTPException(404, detail="Risk report PDF file is missing from storage")

    return StreamingResponse(
        stream,
        media_type="application/pdf",
        headers={"Content-Disposition": f'attachment; filename="risk_report_{contract.name}.pdf"'},
    )


def _get_or_404(db, contract_id: str, user_id: str) -> Contract:
    contract = db.query(Contract).filter(
        Contract.id == contract_id, Contract.owner_id == user_id
    ).first()
    if not contract:
        raise HTTPException(404, detail="Contract not found")
    return contract

File: backend-go/handlers/contracts.go
package handlers

import (
	"archive/zip"
	"bytes"
	"context"
	"encoding/xml"
	"io"
	"io/ioutil"
	"log"
	"path/filepath"
	"strings"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
	"github.com/lexai/backend-go/config"
	"github.com/lexai/backend-go/db"
	"github.com/lexai/backend-go/models"
	"github.com/lexai/backend-go/services"
)

type ContractHandler struct {
	cfg     *config.Config
	storage *services.Storage
	queue   *services.Queue
}

func NewContractHandler(cfg *config.Config, storage *services.Storage, queue *services.Queue) *ContractHandler {
	return &ContractHandler{cfg: cfg, storage: storage, queue: queue}
}

// Upload handles contract file upload, text extraction, and pipeline enqueue.
func (h *ContractHandler) Upload(c *fiber.Ctx) error {
	userID := c.Locals("user_id").(string)

	file, err := c.FormFile("file")
	if err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{"detail": "No file provided"})
	}

	maxBytes := int64(h.cfg.MaxUploadMB) * 1024 * 1024
	if file.Size > maxBytes {
		return c.Status(fiber.StatusRequestEntityTooLarge).JSON(fiber.Map{"detail": "File exceeds size limit"})
	}

	f, err := file.Open()
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{"detail": "Failed to read file"})
	}
	defer f.Close()

	content, err := ioutil.ReadAll(f)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{"detail": "Failed to read file"})
	}

	contractID := uuid.New().String()
	filename := file.Filename

	rawText, err := parseDocument(content, filename)
	if err != nil {
		return c.Status(fiber.StatusUnprocessableEntity).JSON(fiber.Map{"detail": "Could not parse document: " + err.Error()})
	}

	storageKey := "contracts/" + userID + "/" + contractID + "/" + filename
	if err := h.storage.Put(content, storageKey); err != nil {
		log.Printf("[contracts] Storage error: %v", err)
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{"detail": "Could not save uploaded file"})
	}

	ext := strings.ToLower(filepath.Ext(filename))
	if len(ext) > 0 {
		ext = ext[1:]
	}

	now := time.Now().UTC()
	_, err = db.Pool.Exec(context.Background(),
		`INSERT INTO contracts (id, owner_id, name, original_filename, file_type, storage_key, raw_text, clause_count, overall_risk_score, status, thread_id, created_at, updated_at)
		 VALUES ($1, $2, $3, $4, $5, $6, $7, 0, 0, 'pending', '', $8, $9)`,
		contractID, userID, filename, filename, ext, storageKey, rawText, now, now)
	if err != nil {
		log.Printf("[contracts] DB insert error: %v", err)
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{"detail": "Failed to create contract"})
	}

	writeAuditEvent(contractID, "contract_uploaded", &userID, nil, nil,
		map[string]interface{}{"filename": filename, "size_bytes": len(content)})

	job := services.PipelineJob{
		ContractID:   contractID,
		ContractText: rawText,
		ContractName: filename,
		UserID:       userID,
	}
	if err := h.queue.Enqueue(context.Background(), job); err != nil {
		log.Printf("[contracts] Queue error: %v", err)
	}

	return c.Status(fiber.StatusCreated).JSON(models.ContractOut{
		ID: contractID, Name: filename, Status: "pending",
		OverallRiskScore: 0, ClauseCount: 0, ThreadID: "",
	})
}

// List returns all contracts for the authenticated user.
func (h *ContractHandler) List(c *fiber.Ctx) error {
	userID := c.Locals("user_id").(string)
	skip := c.QueryInt("skip", 0)
	limit := c.QueryInt("limit", 20)
	if limit > 100 {
		limit = 100
	}

	rows, err := db.Pool.Query(context.Background(),
		`SELECT id, name, status, overall_risk_score, clause_count, thread_id
		 FROM contracts WHERE owner_id = $1 ORDER BY created_at DESC OFFSET $2 LIMIT $3`,
		userID, skip, limit)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{"detail": "Database error"})
	}
	defer rows.Close()

	contracts := make([]models.ContractOut, 0)
	for rows.Next() {
		var co models.ContractOut
		if err := rows.Scan(&co.ID, &co.Name, &co.Status, &co.OverallRiskScore, &co.ClauseCount, &co.ThreadID); err != nil {
			continue
		}
		contracts = append(contracts, co)
	}
	return c.JSON(contracts)
}

// Get returns a single contract.
func (h *ContractHandler) Get(c *fiber.Ctx) error {
	userID := c.Locals("user_id").(string)
	contractID := c.Params("id")

	var co models.ContractOut
	err := db.Pool.QueryRow(context.Background(),
		`SELECT id, name, status, overall_risk_score, clause_count, thread_id
		 FROM contracts WHERE id = $1 AND owner_id = $2`,
		contractID, userID).Scan(&co.ID, &co.Name, &co.Status, &co.OverallRiskScore, &co.ClauseCount, &co.ThreadID)
	if err != nil {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{"detail": "Contract not found"})
	}
	return c.JSON(co)
}

// DownloadRedline streams the redlined DOCX file.
func (h *ContractHandler) DownloadRedline(c *fiber.Ctx) error {
	userID := c.Locals("user_id").(string)
	contractID := c.Params("id")

	var docxKey, contractName string
	err := db.Pool.QueryRow(context.Background(),
		`SELECT r.redlined_docx_key, c.name FROM reviews r
		 JOIN contracts c ON c.id = r.contract_id
		 WHERE r.contract_id = $1 AND c.owner_id = $2`,
		contractID, userID).Scan(&docxKey, &contractName)
	if err != nil || docxKey == "" {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{"detail": "Redlined DOCX not yet generated"})
	}

	reader, err := h.storage.Get(docxKey)
	if err != nil {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{"detail": "Redlined DOCX file is missing from storage"})
	}
	defer reader.Close()
	data, _ := ioutil.ReadAll(reader)

	c.Set("Content-Type", "application/vnd.openxmlformats-officedocument.wordprocessingml.document")
	c.Set("Content-Disposition", `attachment; filename="redlined_`+contractName+`.docx"`)
	return c.Send(data)
}

// DownloadReport streams the risk report PDF.
func (h *ContractHandler) DownloadReport(c *fiber.Ctx) error {
	userID := c.Locals("user_id").(string)
	contractID := c.Params("id")

	var pdfKey, contractName string
	err := db.Pool.QueryRow(context.Background(),
		`SELECT r.risk_pdf_key, c.name FROM reviews r
		 JOIN contracts c ON c.id = r.contract_id
		 WHERE r.contract_id = $1 AND c.owner_id = $2`,
		contractID, userID).Scan(&pdfKey, &contractName)
	if err != nil || pdfKey == "" {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{"detail": "Risk report PDF not yet generated"})
	}

	reader, err := h.storage.Get(pdfKey)
	if err != nil {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{"detail": "Risk report PDF file is missing from storage"})
	}
	defer reader.Close()
	data, _ := ioutil.ReadAll(reader)

	c.Set("Content-Type", "application/pdf")
	c.Set("Content-Disposition", `attachment; filename="risk_report_`+contractName+`.pdf"`)
	return c.Send(data)
}

// ── Document Parsing ─────────────────────────────────────────────────────────

func parseDocument(content []byte, filename string) (string, error) {
	ext := strings.ToLower(filepath.Ext(filename))
	switch ext {
	case ".txt":
		return string(content), nil
	case ".docx":
		return extractDocxText(content)
	default:
		return string(content), nil
	}
}

func extractDocxText(data []byte) (string, error) {
	reader := bytes.NewReader(data)
	zipReader, err := zip.NewReader(reader, int64(len(data)))
	if err != nil {
		return "", err
	}

	var text strings.Builder
	for _, f := range zipReader.File {
		if f.Name != "word/document.xml" {
			continue
		}
		rc, err := f.Open()
		if err != nil {
			return "", err
		}
		decoder := xml.NewDecoder(rc)
		var inText bool
		for {
			tok, err := decoder.Token()
			if err == io.EOF {
				break
			}
			if err != nil {
				break
			}
			switch t := tok.(type) {
			case xml.StartElement:
				if t.Name.Local == "t" {
					inText = true
				}
			case xml.EndElement:
				if t.Name.Local == "t" {
					inText = false
				}
				if t.Name.Local == "p" {
					text.WriteString("\n")
				}
			case xml.CharData:
				if inText {
					text.Write(t)
				}
			}
		}
		rc.Close()
		break
	}
	return strings.TrimSpace(text.String()), nil
}
